lib/src/repo.rs

```diff
@@ -1154,6 +1154,42 @@ impl MutableRepo {
         )
     }
 
+    /// Rewrite descendants of the given roots.
+    ///
+    /// The callback will be called for each commit with the new parents
+    /// prepopulated. The callback may change the parents and write the new
+    /// commit, or it may abandon the commit, or it may leave the old commit
+    /// unchanged.
+    ///
+    /// The set of commits to visit is determined at the start. If the callback
+    /// adds new descendants, then the callback will not be called for those.
+    /// Similarly, if the callback rewrites unrelated commits, then the callback
+    /// will not be called for descendants of those commits.
+    pub fn transform_descendants(
+        &mut self,
+        settings: &UserSettings,
+        roots: Vec<CommitId>,
+        mut callback: impl FnMut(CommitRewriter) -> BackendResult<()>,
+    ) -> BackendResult<()> {
+        let mut to_visit = self.find_descendants_to_rebase(roots)?;
+        while let Some(old_commit) = to_visit.pop() {
+            let new_parent_ids = self.new_parents(old_commit.parent_ids());
+            let rewriter = CommitRewriter::new(self, old_commit, new_parent_ids);
+            callback(rewriter)?;
+        }
+        self.update_rewritten_references(settings)?;
```

Comment:
I assumed `update_rewritten_references()` is idempotent (so `transform_descendants()` can be called multiple times without clearing the `parent_mapping`), but the assumption might not hold if the working-copy commit got abandoned? Just guess.

Reply:
I have also thought it would be idempotent. What's the case you're thinking of where it might not be? If the working copy pointed to an abandoned commit, it would be moved somewhere else (typically the parents) and would no longer point to a commit in `parent_mapping`, so a subsequent `update_rewritten_references()` should be a no-op. Maybe you're thinking of a more complex case?

Reply:
> Maybe you're thinking of a more complex case?

No, it's just a guess that there might be a problem because `update_wc_commits()` does non-trivial thing. Good to know that the call is idempotent.